Framework/Algorithms/test/MaskInstrumentTest.h

```diff
@@ -0,0 +1,119 @@
+#ifndef MANTID_ALGORITHMS_MASKINSTRUMENTTEST_H_
+#define MANTID_ALGORITHMS_MASKINSTRUMENTTEST_H_
+
+#include <cxxtest/TestSuite.h>
+
+#include "MantidAlgorithms/MaskInstrument.h"
+#include "MantidDataObjects/Workspace2D.h"
+#include "MantidDataObjects/WorkspaceCreation.h"
+#include "MantidAPI/AnalysisDataService.h"
+#include "MantidGeometry/Instrument/DetectorInfo.h"
+
+#include "MantidTestHelpers/InstrumentCreationHelper.h"
+
+using namespace Mantid;
+using namespace Algorithms;
+using namespace API;
+using namespace DataObjects;
+
+namespace {
+MatrixWorkspace_sptr makeWorkspace() {
+  MatrixWorkspace_sptr ws = create<Workspace2D>(4, HistogramData::Points(1));
+  InstrumentCreationHelper::addFullInstrumentToWorkspace(*ws, false, false, "");
+  return ws;
+}
+
+MatrixWorkspace_sptr maskInstrument(const MatrixWorkspace_sptr &ws,
+                                    const std::vector<int> &detectorIDs) {
+  MaskInstrument alg;
+  alg.setRethrows(true);
+  alg.initialize();
+  alg.setProperty("InputWorkspace", ws);
+  alg.setPropertyValue("OutputWorkspace", "out");
+  alg.setProperty("DetectorIDs", detectorIDs);
+  alg.execute();
+  auto out = AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>("out");
+  AnalysisDataService::Instance().remove("out");
+  return out;
+}
+
+MatrixWorkspace_sptr
+maskInstrumentInplace(const MatrixWorkspace_sptr &ws,
+                      const std::vector<int> &detectorIDs) {
+  MaskInstrument alg;
+  alg.setRethrows(true);
+  alg.initialize();
+  alg.setProperty("InputWorkspace", ws);
+  TS_ASSERT_THROWS_NOTHING(
+      alg.setPropertyValue("OutputWorkspace", "_dummy_for_inplace"));
+  alg.setProperty("OutputWorkspace", ws);
+  alg.setProperty("DetectorIDs", detectorIDs);
+  alg.execute();
+  return ws;
+}
+}
+
+class MaskInstrumentTest : public CxxTest::TestSuite {
+public:
+  // This pair of boilerplate methods prevent the suite being created statically
+  // This means the constructor isn't called when running other tests
+  static MaskInstrumentTest *createSuite() { return new MaskInstrumentTest(); }
+  static void destroySuite(MaskInstrumentTest *suite) { delete suite; }
+
+  void test_masking() {
+    const auto in = makeWorkspace();
+    const auto ws = maskInstrument(in, {1, 3});
+    TS_ASSERT_DIFFERS(in, ws);
+    const auto &detInfo = ws->detectorInfo();
+    TS_ASSERT_EQUALS(detInfo.isMasked(0), true);
```

Comment:
A minor point: Might be better to use `indexOf` and work with Detector Ids rather than mixing indexes (for calls to detectorInfo) and ids (used for call to maskInstrument).  At a quick glance I thought that the wrong indices were being masked.

Reply:
I tried and it looked much harder to read. Added a comment at first occurence instead.